AsyncPostgresStore gets stuck in invalid state, requiring full application restart

### Checked other resources

- [x] This is a bug, not a usage question. For questions, please use GitHub Discussions.
- [x] I added a clear and detailed title that summarizes the issue.
- [x] I read what a minimal reproducible example is (https://stackoverflow.com/help/minimal-reproducible-example).
- [x] I included a self-contained, minimal example that demonstrates the issue INCLUDING all the relevant imports. The code run AS IS to reproduce the issue.

### Example Code

To run, checkout this commit https://github.com/darrenclark/langgraph/commit/58a8954d4b7bd51c53c537451b3880ca680021d6 and run:

```sh
cd libs/checkpoint
poetry run pytest tests/test_store.py -k test_async_batch_store_handles_cancellation
```

```python
async def test_async_batch_store_handles_cancellation() -> None:
    class MockStore(AsyncBatchedBaseStore):
        def batch(self, ops: Iterable[Op]) -> list[Result]:
            raise NotImplementedError

        async def abatch(self, ops: Iterable[Op]) -> list[Result]:
            assert all(isinstance(op, GetOp) for op in ops)
            return [
                Item(
                    value={},
                    key=getattr(op, "key", ""),
                    namespace=getattr(op, "namespace", ()),
                    created_at=datetime(2024, 9, 24, 17, 29, 10, 128397),
                    updated_at=datetime(2024, 9, 24, 17, 29, 10, 128397),
                )
                for op in ops
            ]

    store = MockStore()

    # Simulate cancellation
    task = asyncio.create_task(store.aget(namespace=("a",), key="b"))
    await asyncio.sleep(0)
    task.cancel()
    await asyncio.sleep(0)

    # Cancelling individual queries against the store should not break the store
    result = await store.aget(namespace=("c",), key="d")
    assert result == Item(
        value={},
        key="d",
        namespace=("c",),
        created_at=datetime(2024, 9, 24, 17, 29, 10, 128397),
        updated_at=datetime(2024, 9, 24, 17, 29, 10, 128397),
    )
```

### Error Message and Stack Trace (if applicable)

```shell
================================================================================================ test session starts =================================================================================================
platform darwin -- Python 3.11.2, pytest-7.4.4, pluggy-1.5.0 -- /Users/darren/Library/Caches/pypoetry/virtualenvs/langgraph-checkpoint-9BBil99g-py3.11/bin/python
cachedir: .pytest_cache
rootdir: /Users/darren/Projects/langgraph/libs/checkpoint
configfile: pyproject.toml
plugins: anyio-4.9.0, asyncio-0.21.2, mock-3.14.0
asyncio: mode=Mode.AUTO
collected 22 items / 21 deselected / 1 selected

tests/test_store.py::test_async_batch_store_handles_cancellation FAILED

====================================================================================================== FAILURES ======================================================================================================
____________________________________________________________________________________ test_async_batch_store_handles_cancellation _____________________________________________________________________________________

    async def test_async_batch_store_handles_cancellation() -> None:
        class MockStore(AsyncBatchedBaseStore):
            def batch(self, ops: Iterable[Op]) -> list[Result]:
                raise NotImplementedError

            async def abatch(self, ops: Iterable[Op]) -> list[Result]:
                assert all(isinstance(op, GetOp) for op in ops)
                return [
                    Item(
                        value={},
                        key=getattr(op, "key", ""),
                        namespace=getattr(op, "namespace", ()),
                        created_at=datetime(2024, 9, 24, 17, 29, 10, 128397),
                        updated_at=datetime(2024, 9, 24, 17, 29, 10, 128397),
                    )
                    for op in ops
                ]

        store = MockStore()

        # Simulate cancellation
        task = asyncio.create_task(store.aget(namespace=("a",), key="b"))
        await asyncio.sleep(0)
        task.cancel()
        await asyncio.sleep(0)

        # Cancelling individual queries against the store should not break the store
>       result = await store.aget(namespace=("c",), key="d")

tests/test_store.py:184:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <tests.test_store.test_async_batch_store_handles_cancellation.<locals>.MockStore object at 0x1052b4d70>, namespace = ('c',), key = 'd'

    async def aget(
        self,
        namespace: tuple[str, ...],
        key: str,
        *,
        refresh_ttl: Optional[bool] = None,
    ) -> Optional[Item]:
>       assert not self._task.done()
E       AssertionError

langgraph/store/base/batch.py:80: AssertionError
================================================================================================ slowest 5 durations =================================================================================================
0.00s setup    tests/test_store.py::test_async_batch_store_handles_cancellation
0.00s call     tests/test_store.py::test_async_batch_store_handles_cancellation
0.00s teardown tests/test_store.py::test_async_batch_store_handles_cancellation
============================================================================================== short test summary info ===============================================================================================
FAILED tests/test_store.py::test_async_batch_store_handles_cancellation - AssertionError
========================================================================================== 1 failed, 21 deselected in 0.11s ==========================================================================================
```

### Description

* When a query against `AsyncPostgresStore` gets cancelled, it gets in a broken state
    * For us, _I think_ this cancellation happens indirectly via gRPC Python when we hit a deadline.
    * ‼️ **Specifically, `AsyncBatchedBaseStore`'s `_task` dies & never gets restarted** 
* Once this happens, the `AsyncPostgresStore` is stuck in a bad state.
    * To repair, the app needs to be shutdown & restarted

I've included a failing test case here that distills the issue: https://github.com/darrenclark/langgraph/commit/58a8954d4b7bd51c53c537451b3880ca680021d6

The root cause of the issue seems to be in this block of code: https://github.com/langchain-ai/langgraph/blob/2c2ace2a404602b913f4de3dc74d251fe188edc7/libs/checkpoint/langgraph/store/base/batch.py#L345-L349

I'm sure there are different code paths that trigger this, but for us I saw:

1. `set_result` raises a `asyncio.exception.InvalidStateError` since it's cancelled
2. that `InvalidStateError` error gets caught by the `except Exception as e` block
3. `set_exception` (inside except block) then also raises a `asyncio.exception.InvalidStateError`
4. that `asyncio.exception.InvalidStateError` bubbles out & kills the batching loop task
5. future queries to the store fail on `assert not self._task.done()`

### System Info


System Information
------------------
> OS:  Darwin
> OS Version:  Darwin Kernel Version 24.5.0: Tue Apr 22 19:53:27 PDT 2025; root:xnu-11417.121.6~2/RELEASE_ARM64_T6041
> Python Version:  3.11.2 (main, Nov 23 2024, 16:33:44) [Clang 16.0.0 (clang-1600.0.26.4)]

Package Information
-------------------
> langchain_core: 0.3.68
> langsmith: 0.4.5

Optional packages not installed
-------------------------------
> langserve

Other Dependencies
------------------
> httpx: 0.28.1
> jsonpatch<2.0,>=1.33: Installed. No version info available.
> langsmith-pyo3: Installed. No version info available.
> langsmith>=0.3.45: Installed. No version info available.
> openai-agents: Installed. No version info available.
> opentelemetry-api: Installed. No version info available.
> opentelemetry-exporter-otlp-proto-http: Installed. No version info available.
> opentelemetry-sdk: Installed. No version info available.
> orjson: 3.10.18
> packaging: 24.2
> packaging<25,>=23.2: Installed. No version info available.
> pydantic: 2.10.2
> pydantic>=2.7.4: Installed. No version info available.
> pytest: 8.3.3
> PyYAML>=5.3: Installed. No version info available.
> requests: 2.32.3
> requests-toolbelt: 1.0.0
> rich: 13.9.4
> tenacity!=8.4.0,<10.0.0,>=8.1.0: Installed. No version info available.
> typing-extensions>=4.7: Installed. No version info available.
> zstandard: 0.23.0

--- a/libs/checkpoint/langgraph/store/base/batch.py
+++ b/libs/checkpoint/langgraph/store/base/batch.py
@@ -343,10 +343,14 @@ async def _run(
 
                     # set the results of each operation
                     for fut, result in zip(futs, results):
-                        fut.set_result(result)
+                        # guard against future being done (e.g. cancelled)
+                        if not fut.done():
+                            fut.set_result(result)
                 except Exception as e:
                     for fut in futs:
-                        fut.set_exception(e)
+                        # guard against future being done (e.g. cancelled)
+                        if not fut.done():
+                            fut.set_exception(e)
             finally:
                 # remove strong ref to store
                 del s

--- a/libs/checkpoint/tests/test_store.py
+++ b/libs/checkpoint/tests/test_store.py
@@ -155,6 +155,43 @@ async def test_async_batch_store(mocker: MockerFixture) -> None:
     ]
 
 
+async def test_async_batch_store_handles_cancellation() -> None:
+    class MockStore(AsyncBatchedBaseStore):
+        def batch(self, ops: Iterable[Op]) -> list[Result]:
+            raise NotImplementedError
+
+        async def abatch(self, ops: Iterable[Op]) -> list[Result]:
+            assert all(isinstance(op, GetOp) for op in ops)
+            return [
+                Item(
+                    value={},
+                    key=getattr(op, "key", ""),
+                    namespace=getattr(op, "namespace", ()),
+                    created_at=datetime(2024, 9, 24, 17, 29, 10, 128397),
+                    updated_at=datetime(2024, 9, 24, 17, 29, 10, 128397),
+                )
+                for op in ops
+            ]
+
+    store = MockStore()
+
+    # Simulate cancellation
+    task = asyncio.create_task(store.aget(namespace=("a",), key="b"))
+    await asyncio.sleep(0)
+    task.cancel()
+    await asyncio.sleep(0)
+
+    # Cancelling individual queries against the store should not break the store
+    result = await store.aget(namespace=("c",), key="d")
+    assert result == Item(
+        value={},
+        key="d",
+        namespace=("c",),
+        created_at=datetime(2024, 9, 24, 17, 29, 10, 128397),
+        updated_at=datetime(2024, 9, 24, 17, 29, 10, 128397),
+    )
+
+
 def test_list_namespaces_basic() -> None:
     store = InMemoryStore()
 

